pkg/endpoint/policy.go

```diff
@@ -711,6 +720,22 @@ func (e *Endpoint) Regenerate(regenMetadata *regeneration.ExternalRegenerationMe
 	}
 
 	regenContext := ParseExternalRegenerationMetadata(ctx, cFunc, regenMetadata)
+	stats := &regenContext.Stats
+	datapathRegenCtxt := regenContext.datapathRegenerationContext
+
+	// Compute Endpoint's policy before queueing the regeneration event.
+	stats.policyCalculation.Start()
+	err := e.regeneratePolicy(stats, datapathRegenCtxt)
+	stats.policyCalculation.End(err == nil)
+	if err != nil {
+		cFunc()
+		if !errors.Is(err, context.Canceled) {
+			e.getLogger().WithError(err).Error("unable to regenerate policy")
```

Comment:
What is the consequence if the policy is not regenerated?

Reply:
This is a very good question!

Upon reflecting on this I decided to add a 5th commit to clean this up as follows:
- separate the "policy precomputation" code to a new function `ComputeInitialPolicy()` to minimize added complexity in `Regenerate()`
- `ComputeInitialPolicy()` only fails if the endpoint can not be locked. In all other failure cases it lets the regeneration be queued normally, so that anything that failed in `ComputeInitialPolicy` can be tried again in the normal code path
- The `InitialEnvoyPolicyComputed` channel can now be closed from two different places, created a helper function `InitialPolicyComputedLocked()` for it to be safe (endpoint must be locked, never close the channel twice).

Reply:
@doniacld please have a 2nd look!

Reply:
Thanks for your explanations! I find the new path much easier to read and understand especially the `other failure cases` part.